src/rust/bitbox02-rust/src/backup.rs

```diff
@@ -0,0 +1,357 @@
+
+use crate::pb_backup;
+
+use alloc::boxed::Box;
+use alloc::string::String;
+use alloc::vec::Vec;
+
+use hmac::{Hmac, Mac, NewMac};
+use sha2::{Digest, Sha256};
+use zeroize::{Zeroize, Zeroizing};
+
+#[derive(Debug)]
+pub enum Error {
+    Generic,
+    Stale,
+    SdList,
+    SdRead,
+    SdWrite,
+    Check,
+}
+
+#[derive(Default)]
+pub struct BackupData(pub Box<pb_backup::BackupData>);
+
+impl BackupData {
+    /// The seed field is always 32 bytes, a historical "accident" as it was easier to fix the size
+    /// when using C (and nanopb protobuf), with the seed_length indicating the actual length.
+    pub fn get_seed(&self) -> &[u8] {
+        &self.0.seed[..self.0.seed_length as _]
+    }
+}
+
+impl Zeroize for BackupData {
+    fn zeroize(&mut self) {
+        self.0.seed_length.zeroize();
+        self.0.seed.zeroize();
+        self.0.birthdate.zeroize();
+        self.0.generator.zeroize();
+    }
+}
+/// Pad a seed with zeroes to the right up to 32 bytes.  That the seed field in backups is always 32
+/// bytes is a historical "accident" as it was easier to fix the size when using C (and nanopb
+/// protobuf), with the seed_length indicating the actual length.
+fn padded_seed(seed: &[u8]) -> Zeroizing<[u8; 32]> {
+    let mut result = Zeroizing::new([0u8; 32]);
+    result[..seed.len()].copy_from_slice(seed);
+    result
+}
+
+pub fn id(seed: &[u8]) -> String {
+    let mut mac = Hmac::<Sha256>::new_from_slice(b"backup").unwrap();
+    mac.update(padded_seed(seed).as_ref());
+    hex::encode(mac.finalize_reset().into_bytes())
+}
+
+fn compute_checksum(
+    metadata: &pb_backup::BackupMetaData,
+    data: &pb_backup::BackupData,
+    data_length: u32,
+) -> Result<Vec<u8>, ()> {
+    let mut hasher = Sha256::new();
+    hasher.update(metadata.timestamp.to_le_bytes());
+    let mode: u8 = metadata.mode as _;
+    hasher.update(mode.to_le_bytes());
+    let padded_name: [u8; 64] = {
+        let mut result = [0u8; 64];
+        let name = metadata.name.as_bytes();
+        if name.len() > 64 {
+            return Err(());
+        }
+        result[..name.len()].copy_from_slice(name);
+        result
+    };
+    hasher.update(&padded_name);
+    hasher.update(data.seed_length.to_le_bytes());
+    if data.seed.len() != 32 {
+        return Err(());
+    }
+    hasher.update(&padded_seed(&data.seed));
+    hasher.update(data.birthdate.to_le_bytes());
+    let padded_generator: [u8; 20] = {
+        let mut result = [0u8; 20];
+        let generator = data.generator.as_bytes();
+        if generator.len() > 20 {
+            return Err(());
+        }
+        result[..generator.len()].copy_from_slice(generator);
+        result
+    };
+    hasher.update(&padded_generator);
+    hasher.update(data_length.to_le_bytes());
+    Ok(hasher.finalize().to_vec())
+}
+
+fn load_from_buffer(buf: &[u8]) -> Result<(Zeroizing<BackupData>, pb_backup::BackupMetaData), ()> {
+    let backup = pb_backup::Backup::decode(buf).or(Err(()))?;
```

Comment:
need to check if the name and generator were serialized with fixed 20/64 bytes (zero padded) by C (null terminated strings) and what consequences this could have here